pkg/nsx/services/subnetport/subnetport.go

```diff
@@ -0,0 +1,188 @@
+	SubnetPortStore *SubnetPortStore
+}
+
+// InitializeSubnetPort sync NSX resources
+func InitializeSubnetPort(service common.Service) (*SubnetPortService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+
+	subnetPortService := &SubnetPortService{Service: service}
+
+	subnetPortService.SubnetPortStore = &SubnetPortStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeSubnetPortCRUID: indexFunc}),
+		BindingType: model.VpcBindingType(),
+	}}
+
+	go subnetPortService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnetPort, subnetPortService.SubnetPortStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetPortService, err
+	}
+
+	return subnetPortService, nil
+}
+
+func (service *SubnetPortService) CreateOrUpdateSubnetPort(obj *v1alpha1.SubnetPort) error {
+
+	log.Info("Create or update subnetport", "subnetport", obj)
+
+	nsxSubnetPort, err := service.buildSubnetPort(obj)
+	if err != nil {
+		log.Error(err, "failed to build NSX subnet port")
+		return err
+	}
+
+	// TODO: For now, only the parent subnet is supported.
+	specSubnet := obj.Spec.Subnet
+
+	// TODO: Get the subnet's NSX resource path from subnet store after the store is implemented.
+	nsxSubnetPath, err := service.GetPathFromSubnet(specSubnet, obj.Namespace, service.Client)
+	if err != nil {
+		log.Error(err, "failed to get NSX resource path from subnet", "subnet", specSubnet, "namespace", obj.Namespace)
+	}
+
+	paras := strings.Split(nsxSubnetPath, "/")
+	nsxOrgID := paras[2]
+	nsxProjectID := paras[4]
+	nsxVPCID := paras[6]
+	nsxSubnetID := paras[8]
+
+	existingSubnetPort := service.SubnetPortStore.GetByKey(*nsxSubnetPort.Id)
+	isChanged := common.CompareResource(SubnetPortToComparable(existingSubnetPort), SubnetPortToComparable(nsxSubnetPort))
+	if !isChanged {
+		log.Info("NSX subnet port not changed, skipping the update", "nsxSubnetPort.Id", nsxSubnetPort.Id)
+		return nil
+	}
+	err = service.NSXClient.PortClient.Patch(nsxOrgID, nsxProjectID, nsxVPCID, nsxSubnetID, *nsxSubnetPort.Id, *nsxSubnetPort)
+	if err != nil {
+		log.Error(err, "failed to create or update subnet port")
+		return err
+	}
+	err = service.SubnetPortStore.Operate(nsxSubnetPort)
+	if err != nil {
+		return err
+	}
+	log.Info("successfully updated NSX subnet port", "subnetPort", nsxSubnetPort)
+	return nil
+}
+
+func (service *SubnetPortService) GetPathFromSubnet(subnetName string, subnetNamespace string, client client.Client) (string, error) {
+	path := ""
+	subnetList := &v1alpha1.SubnetList{}
+	err := client.List(context.Background(), subnetList)
+	if err != nil {
+		log.Error(err, "failed to list all the subnet CRs")
+		return path, err
+	}
+	for _, subnet := range subnetList.Items {
+		if subnet.Name == subnetName {
+			path = subnet.Status.NSXResourcePath
```

Comment:
can we get Subnet object and then get the subnet path from status? why we list all subnets here and compare with name?

Reply:
done, thanks